apps/events/models.py

```diff
@@ -0,0 +1,165 @@
+    action = models.OneToOneField(EventAction, on_delete=models.CASCADE, related_name="timeout_trigger")
+    experiment = models.ForeignKey(Experiment, on_delete=models.CASCADE, related_name="timeout_triggers")
+    delay = models.PositiveIntegerField(
+        help_text="The amount of time in seconds to fire this trigger.",
+    )
+    total_num_triggers = models.IntegerField(
+        default=1,
+        help_text="The number of times to fire this trigger",
+    )
+    event_logs = GenericRelation(EventLog)
+
+    def timed_out_sessions(self):
+        """Finds all the timed out sessions where:
+        - The last human message was sent at a time earlier than the trigger time
+        - There have been fewer trigger attempts than the total number defined by the trigger
+        """
+
+        trigger_time = timezone.now() - timedelta(seconds=self.delay)
+
+        last_human_message_created_at = (
+            ChatMessage.objects.filter(
+                chat__experiment_session=OuterRef("pk"),
+                message_type=ChatMessageType.HUMAN,
+            )
+            .order_by("-created_at")
+            .values("created_at")[:1]
+        )
+        last_human_message_id = (
+            ChatMessage.objects.filter(
+                chat__experiment_session=OuterRef("session_id"),
+                message_type=ChatMessageType.HUMAN,
+            )
+            .order_by("-created_at")
+            .values("id")[:1]
+        )
+        log_count_for_last_message = (
+            EventLog.objects.filter(
+                session=OuterRef("pk"),
+                chat_message_id=Subquery(last_human_message_id),
+                status=EventLogStatusChoices.SUCCESS,
+            )
+            .annotate(
+                count=Func(F("chat_message_id"), function="Count")
+            )  # We don't use Count here because otherwise Django wants to do a group_by, which messes up the subquery: https://stackoverflow.com/a/69031027
+            .values("count")
+        )
+
+        sessions = (
+            ExperimentSession.objects.filter(
+                experiment=self.experiment,
+                ended_at=None,
+            )
+            .annotate(
+                last_human_message_created_at=Subquery(last_human_message_created_at),
+                log_count=Subquery(log_count_for_last_message),
+            )
+            .filter(
+                last_human_message_created_at__lt=trigger_time,
+                last_human_message_created_at__isnull=False,
+            )  # The last message was received before the trigger time
+            .filter(
+                Q(log_count__lt=self.total_num_triggers) | Q(log_count__isnull=True)
+            )  # There were either no tries yet, or fewer tries than the required number for this message
+        )
+        return sessions.all()
+
+    def fire(self, session):
+        last_human_message = session.chat.messages.filter(message_type=ChatMessageType.HUMAN).last()
```

Comment:
I haven't checked where this get's called but doing these long chains can lead to a bunch of extra DB queries. It might be better to just query `ChatMessage` directly using `session.chat_id`

Reply:
Great point. I'll update it.